### PR TITLE
Added Gson serialization/deserialization Extension

### PR DESCRIPTION
**Pull request by patton73**
I have provided Gson Extension to serialize/deserialize json.
I have also provided tests as in the other extensions.

### CODE REVIEW

**Comment by lhazlewood** [comment]
Thank's for the PR!  We'll see if we can get this in the next minor release!

**Comment by romeara** [comment]
This is something I'd be interested in using - if further work is needed before this is ready to be added, please let me know and I'd be happy to help!

**Comment by DanFTRX** [comment]
I think you need to actually add it to the locator: https://github.com/jwtk/jjwt/blob/master/impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathSerializerLocator.java#L32 

And at that point you would probably want https://github.com/jwtk/jjwt/pull/425  to come in first because GSON is a popular library and the current locator method would then be choosing a different serializer without people knowing it.

**Comment by patton73** [comment]
Is there a particular reason that keeps this pull request blocked for about 3 months? And more if https://github.com/jwtk/jjwt/pull/425 is a precondition why this is not included in 0.11.0 milestone?
Thanks.

**Comment by lhazlewood** [comment]
@patton73 no reason other than it has been a lower priority than shoring up the JWE work.  I hope I'll be able to get to this this week.

**Comment by lhazlewood** [comment]
P.S. *Thank you* for the PR!  It has just been crazy lately, so hopefully we can address this asap.

**Comment by DanFTRX** [comment]
@patton73 This still needs to add the relevant lines to https://github.com/jwtk/jjwt/blob/master/impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathSerializerLocator.java  to load the GSON extension when present.

**Comment by patton73** [comment]
> @patton73 This still needs to add the relevant lines to https://github.com/jwtk/jjwt/blob/master/impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathSerializerLocator.java to load the GSON extension when present.

Yes i know. i Thought they could merge this PR earlier. I think it can be done by mantainers. It is only a matter of add another if with right class and package.

**Comment by lhazlewood** [comment]
One significant reason this PR hasn't been merged yet is that the RuntimeClasspathSerializerLocator hasn't been updated to load GSON when present.  Is there a reason why you didn't want to update that?

File: extensions/pom.xml [hunk 1]
```diff
@@ -36,6 +36,7 @@
     <modules>
         <module>jackson</module>
         <module>orgjson</module>
+	<module>gson</module>
```
Comment (lhazlewood) [inline extensions/pom.xml]: Minor formatting fix request.

**Comment by patton73** [comment]
I did not understand this PR was waiting those changes. I will update i as soon as i can.

**Comment by patton73** [comment]
Mmm. It did not compile. it complaints about "Could not find artifact io.jsonwebtoken:jjwt-root:pom:0.11.0-SNAPSHOT in sonatype-snapshots". What should i do?

**Comment by lhazlewood** [comment]
@patton73 I'm digging in to this today - were you able to rebase from master?

**Comment by patton73** [comment]
I did not have problems in rebasing the master.

**Comment by lhazlewood** [comment]
@patton73 but the build is failing: https://travis-ci.org/jwtk/jjwt/builds/559075398?utm_source=github_status&utm_medium=notification

**Comment by patton73** [comment]
Same strange error about sonatype artifacts. I am afraid I do not know what to do.

**Comment by DanFTRX** [comment]
@patton73 @lhazlewood Probably because your pom references [0.11.0-SNAPSHOT](https://github.com/jwtk/jjwt/pull/414/files#diff-905f6582f56a9b724aa26f8467672a09R24) but the root pom in the repo is [0.10.8-SNAPSHOT](https://github.com/jwtk/jjwt/blob/master/extensions/orgjson/pom.xml#L24)

**Comment by patton73** [comment]
I try to modify root pom in the repo to 0.10.8-SNAPSHOT and see what happens.

**Comment by patton73** [comment]
Ok it worked. It still complaints in the jdk8 compilation but the problem is solved.

File: impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java [hunk 2]
```diff
@@ -34,6 +34,8 @@
             return Classes.newInstance("io.jsonwebtoken.io.JacksonDeserializer");
         } else if (isAvailable("io.jsonwebtoken.io.OrgJsonDeserializer")) {
             return Classes.newInstance("io.jsonwebtoken.io.OrgJsonDeserializer");
+        } else if (isAvailable("com.google.gson.GsonBuilder")) {
```
Comment (DanFTRX) [inline impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java]: This should be checking for "io.jsonwebtoken.io.GsonDeserializer" not "com.google.gson.GsonBuilder"

File: impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathSerializerLocator.java [hunk 3]
```diff
@@ -34,6 +34,8 @@
             return Classes.newInstance("io.jsonwebtoken.io.JacksonSerializer");
         } else if (isAvailable("io.jsonwebtoken.io.OrgJsonSerializer")) {
             return Classes.newInstance("io.jsonwebtoken.io.OrgJsonSerializer");
+        } else if (isAvailable("com.google.gson.GsonBuilder")) {
```
Comment (DanFTRX) [inline impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathSerializerLocator.java]: This should be checking for "io.jsonwebtoken.io.GsonDeserializer" not "com.google.gson.GsonBuilder"

File: impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java [hunk 4]
```diff
@@ -34,6 +34,8 @@
...
+            return Classes.newInstance("io.jsonwebtoken.io.GsonDeSerializer");
```
Comment (DanFTRX) [inline impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java]: This should be "io.jsonwebtoken.io.GsonDeserializer" not "io.jsonwebtoken.io.GsonDeSerializer"

File: impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathSerializerLocator.java (hunk 3 above)
Comment (patton73) [reply impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathSerializerLocator.java]: Ok patching.

File: impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java (hunk 4 above)
Comment (patton73) [reply impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java]: Ok patching.

**Comment by DanFTRX** [comment]
@patton73 I think this should do it:  https://github.com/DanFTRX/jjwt/commit/4c89f0263797eec03a8751119b1b4879b2de9c76

File: impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java (hunk 2 above)
Comment (patton73) [reply impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java]: ok fixed.

File: pom.xml [hunk 5]
```diff
@@ -69,7 +69,7 @@
         <pluginRepository>
             <snapshots>
                 <enabled>false</enabled>
-            </snapshots>
+           </snapshots>
```
Comment (DanFTRX) [inline pom.xml]: Formatting changed?

File: impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java (hunk 4 above)
Comment (DanFTRX) [reply impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java]: Missed DeSerializer -> Deserializer on both this and the line above now.

File: pom.xml (hunk 5 above)
Comment (patton73) [reply pom.xml]: Mmm it does not find io.jsonwebtoken.io.GsonSerializer in the groovy test.

Comment (patton73) [reply pom.xml]: Anyway the typos problem are resolved now.

**Comment by DanFTRX** [comment]
@patton73 I missed a pom, you need to add the gson extension to this:  https://github.com/jwtk/jjwt/blob/master/impl/pom.xml  as a test dependency

File: impl/pom.xml [hunk 6]
```diff
@@ -54,6 +54,11 @@
             <artifactId>jjwt-orgjson</artifactId>
             <scope>test</scope>
         </dependency>
+	<dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-gson</artifactId>
+	   <version>${project.version}</version>
```
Comment (DanFTRX) [inline impl/pom.xml]: Not quite.  Drop the version tag, and add a test scope like the one above it, and make sure the formatting is correct.

Comment (patton73) [reply impl/pom.xml]: Excuse me i ude gradle basically. Whats wrong with farmatting? do i need to use tabs?

Comment (DanFTRX) [reply impl/pom.xml]: You have less spaces than the line above it. so it is misaligned.

File: pom.xml (hunk 5 above)
Comment (DanFTRX) [inline pom.xml]: Oof, it looks like you forgot to revert the change to this line.  You accidentally removed one of the spaces..

**Comment by DanFTRX** [comment]
@lhazlewood Aside from the one formatting change, I think its presentable now.  I'm not sure whats up with the JDK9 travis failure, but it doesn't appear to be caused by the changes.

File: pom.xml (hunk 5 above)
Comment (patton73) [reply pom.xml]: Hopefully fixed.

**Comment by DanFTRX** [comment]
@patton73 Looks good to me, now its up to @lhazlewood

File: extensions/pom.xml (hunk 1 above)
Comment (DanFTRX) [reply extensions/pom.xml]: this is a tab, needs be be 8 spaces instead

File: impl/pom.xml [hunk 7]
```diff
@@ -54,6 +54,11 @@
             <artifactId>jjwt-orgjson</artifactId>
             <scope>test</scope>
         </dependency>
+	<dependency>
```
Comment (DanFTRX) [inline impl/pom.xml]: This is also a tab, change to 8 spaces

File: impl/pom.xml [hunk 8]
```diff
@@ -54,6 +54,11 @@
...
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-gson</artifactId>
+	    <scope>test</scope>
```
Comment (DanFTRX) [inline impl/pom.xml]: This has a tab and spaces,  make sure it is 12 spaces, no tabs

File: impl/pom.xml [hunk 9]
```diff
@@ -54,6 +54,11 @@
...
+	</dependency>
```
Comment (DanFTRX) [inline impl/pom.xml]: This is a tab, change to 8 spaces

File: pom.xml [hunk 10]
```diff
@@ -131,6 +132,11 @@
                 <artifactId>jjwt-orgjson</artifactId>
                 <version>${project.version}</version>
             </dependency>
+	    <dependency>
```
Comment (DanFTRX) [inline pom.xml]: this has a tab and spaces, make sure it is 12 spaces, no tabs

File: pom.xml [hunk 11]
```diff
@@ -90,6 +90,7 @@
 
         <jackson.version>2.9.9.1</jackson.version>
         <orgjson.version>20180130</orgjson.version>
+	<gson.version>2.8.5</gson.version>
```
Comment (DanFTRX) [inline pom.xml]: This is a tab, change to 8 spaces

File: pom.xml [hunk 12]
```diff
@@ -141,6 +147,11 @@
                 <artifactId>json</artifactId>
                 <version>${orgjson.version}</version>
             </dependency>
+	    <dependency>
```
Comment (DanFTRX) [inline pom.xml]: this has a tab and spaces, make sure it is 12 spaces, no tabs

File: extensions/gson/src/main/java/io/jsonwebtoken/io/GsonDeserializer.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
+package io.jsonwebtoken.io;
```
Comment (lhazlewood) [inline extensions/gson/src/main/java/io/jsonwebtoken/io/GsonDeserializer.java]: Because of the Java >= 9 module system, separate implementation .jars cannot export the same packages as the base api .jar.  As a result we have to fix this for the Jackson and org.json implementations and that's being tracked under #399 .

But because this gson extension is new code, we can avoid this mistake and do it properly from the start.  As a result, please move these to a `io.jsonwebtoken.gson.io` package.

This is to use a convention moving forward:

`io.jsonwebtoken` prefix + `.<extension-name>.` + `<mirrored suffix in the api module>`

For example, if there is an interface in the API module in the `io.jsonwebtoken.foo` package, and an extension named `bar` wanted to implement it, then:

* the bar implementation would be in package `io.jsonwebtoken.bar.foo`, and
* the .java file would be in `extensions/bar/src/main/java/io/jsonwebtoken/bar/foo`

This works for the Java >= 9 module system because `io.jsonwebtoken.*` is exported from the api .jar, and only `io.jsonwebtoken.gson.*` is exported from the gson .jar.  There is no set intersection of packages between the two (which the JDK Module System requires).

File: extensions/gson/src/main/java/io/jsonwebtoken/io/GsonSerializer.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
+package io.jsonwebtoken.io;
```
Comment (lhazlewood) [inline extensions/gson/src/main/java/io/jsonwebtoken/io/GsonSerializer.java]: Per the same explanation of `GsonDeserializer`, please move this to `io.jsonwebtoken.gson.io`

File: extensions/gson/src/main/java/io/jsonwebtoken/io/GsonSerializer.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    public GsonSerializer(Gson gson) {
+        Assert.notNull(gson, "gson cannot be null.");
+        this.gson = gson;
+    }
+
+    @Override
+    public byte[] serialize(T t) throws SerializationException {
+        Assert.notNull(t, "Object to serialize cannot be null.");
+        //Gson never throws any serialization exception 
+        return writeValueAsBytes(t);
```
Comment (lhazlewood) [inline extensions/gson/src/main/java/io/jsonwebtoken/io/GsonSerializer.java]: Per the other implementations (Jackson and org.json), this call is wrapped in a try/catch block, and if any `Exception` is thrown, it is wrapped and re-thrown as a `io.jsonwebtoken.io.SerializationException`.  To ensure runtime consistency, we need to do that here too, i.e.

```java
@Override
public byte[] serialize(T t) throws SerializationException {
    Assert.notNull(t, "Object to serialize cannot be null.");
    try {
        return writeValueAsBytes(t);
    } catch (Exception e) {
        String msg = "Unable to serialize object: " + e.getMessage();
        throw new SerializationException(msg, e);
    }
}
```

File: extensions/gson/src/test/groovy/io/jsonwebtoken/io/GsonDeserializerTest.groovy [hunk 16]
```diff
@@ -0,0 +1,64 @@
+package io.jsonwebtoken.io
```
Comment (lhazlewood) [inline extensions/gson/src/test/groovy/io/jsonwebtoken/io/GsonDeserializerTest.groovy]: `io.jsonwebtoken.gson.io`

File: extensions/gson/src/test/groovy/io/jsonwebtoken/io/GsonSerializerTest.groovy [hunk 17]
```diff
@@ -0,0 +1,99 @@
+package io.jsonwebtoken.io
```
Comment (lhazlewood) [inline extensions/gson/src/test/groovy/io/jsonwebtoken/io/GsonSerializerTest.groovy]: `io.jsonwebtoken.gson.io`

File: impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java [hunk 18]
```diff
@@ -34,6 +34,8 @@
             return Classes.newInstance("io.jsonwebtoken.io.JacksonDeserializer");
         } else if (isAvailable("io.jsonwebtoken.io.OrgJsonDeserializer")) {
             return Classes.newInstance("io.jsonwebtoken.io.OrgJsonDeserializer");
+        } else if (isAvailable("io.jsonwebtoken.io.GsonDeserializer")) {
```
Comment (lhazlewood) [inline impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathDeserializerLocator.java]: `io.jsonwebtoken.gson.io.GsonDeserializer`

File: impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathSerializerLocator.java [hunk 19]
```diff
@@ -34,6 +34,8 @@
             return Classes.newInstance("io.jsonwebtoken.io.JacksonSerializer");
         } else if (isAvailable("io.jsonwebtoken.io.OrgJsonSerializer")) {
             return Classes.newInstance("io.jsonwebtoken.io.OrgJsonSerializer");
+        } else if (isAvailable("io.jsonwebtoken.io.GsonSerializer")) {
```
Comment (lhazlewood) [inline impl/src/main/java/io/jsonwebtoken/impl/io/RuntimeClasspathSerializerLocator.java]: `io.jsonwebtoken.gson.io.GsonSerializer`

**Comment by lhazlewood** [comment]
@patton73 will you be able to get to this today?  Thoughts?

**Comment by patton73** [comment]
Pheraps this night. Or at max tomorrow morning. I am travelling and no laptop with me.

**Comment by lhazlewood** [comment]
@patton73 understood - enjoy your trip! Thanks!

**Comment by DanFTRX** [comment]
Whitespace checks out.

File: extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Test
+    void testSerialize() {
+        byte[] expected = '{"hello":"世界"}'.getBytes(Strings.UTF_8)
+        byte[] result = new GsonSerializer().serialize([hello: '世界'])
+        assertTrue Arrays.equals(expected, result)
+    }
+
+//THIS IS COMMENTED OUT BECAUSE GSN NEVER THROWS ANY SERIALIZATION EXCEPTION
```
Comment (DanFTRX) [inline extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy]: You have to re-enable this test @patton73 Code coverage requires it.

Comment (patton73) [reply extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy]: Well gson does not lauch any exception do I write a fake test about that?

Comment (DanFTRX) [reply extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy]: The test actually handles that.  It throws an exception by overriding the serialization on line 82

Comment (DanFTRX) [reply extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy]: Just choose a different exception.  ANy would do.  the OrgJson extension uses IllegalArgumentException:  https://github.com/jwtk/jjwt/blob/73f7f9915baccaa3f3143e549647f818606d49ce/extensions/orgjson/src/test/groovy/io/jsonwebtoken/io/OrgJsonSerializerTest.groovy#L35

Comment (patton73) [reply extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy]: If i re-enable this the test will never pass since gson will never throw a serialization exception.

Comment (DanFTRX) [reply extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy]: The following code creates a Mock GSONSerializer that DOES throw an exception.  Now I agree its a bit silly because GSON won't even throw an exception outside of this unit test, but @lhazlewood wanted the try catch there for consistency and to maintain 100% code coverage you need to hit that try catch
```
    @Test
    void testSerializeFailsWithWrappedException() {

        def ex = createMock(Exception)

        expect(ex.getMessage()).andReturn('foo')

        def serializer = new GsonSerializer() {
            @Override
            protected byte[] writeValueAsBytes(Object o) {
                throw ex
            }
        }

        replay ex

        try {
            serializer.serialize([hello: 'world'])
            fail()
        } catch (SerializationException se) {
            assertEquals 'Unable to serialize object: foo', se.getMessage()
            assertSame ex, se.getCause()
        }

        verify ex
    }```

Comment (patton73) [reply extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy]: Just for curiosity, is there a reason behind the use of groovy to test java instead of junit?

**Comment by DanFTRX** [comment]
@lhazlewood No problem, happy to help.

File: extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy (hunk 20 above)
Comment (lhazlewood) [reply extensions/gson/src/test/groovy/io/jsonwebtoken/gson/io/GsonSerializerTest.groovy]: @patton73 we are using JUnit as the testing library.  We just use Groovy to write the tests - it's faster to write, allows us to ignore Java visibility modifiers to assert values and state, reflection is far easier, mocking things can be simpler, and it's fun to use.

In other words, it's far simpler and more productive to write tests in Groovy, but we still use JUnit to execute the tests and assertions.  HTH!
